# The University of Hong Kong
## DASC7600 Data Science Project 2024
## EDA - HK - Case Counts

# Import Modules and Settings

In [1]:
import os
import sys

# Add project directory to system path
project_dir = os.path.dirname(os.getcwd())
sys.path.append(project_dir)

In [2]:
import matplotlib.pyplot as plt
import pandas as pd
import warnings

import covid_module

# Settings
pd.set_option('display.max_columns', None)
warnings.filterwarnings('ignore')

%matplotlib inline

# Load Data

In [3]:
# Read csv file
## Dataset name: Data in Coronavirus Disease (COVID-19)
## Data resource name: Latest situation of reported cases of COVID-19 in Hong Kong (English)
## URL: https://data.gov.hk/en-data/dataset/hk-dh-chpsebcddr-novel-infectious-agent/resource/9252c845-3aea-4ea7-abae-b385916106b3
covid_hk_case_count_df = pd.read_csv(project_dir + '/data/raw_data/hk/latest_situation_of_reported_cases_covid_19_eng.csv')

# Basic Information of Data Set

In [4]:
# Print first 10 records
covid_hk_case_count_df.head(10)

As of date As of time  Number of confirmed cases  \
0  08/01/2020      12:00                        0.0   
1  09/01/2020      12:00                        0.0   
2  10/01/2020      12:00                        0.0   
3  11/01/2020      12:00                        0.0   
4  12/01/2020      12:00                        0.0   
5  13/01/2020      12:00                        0.0   
6  14/01/2020      12:00                        0.0   
7  15/01/2020      12:00                        0.0   
8  16/01/2020      12:00                        0.0   
9  17/01/2020      12:00                        0.0   

   Number of ruled out cases  \
0                       21.0   
1                       25.0   
2                       31.0   
3                       46.0   
4                       51.0   
5                       56.0   
6                       60.0   
7                       66.0   
8                       71.0   
9                       75.0   

   Number of cases still hospitalised for investigation  \
0                                               17.0      
1                                               23.0      
2                                               23.0      
3                                               15.0      
4                                               16.0      
5                                               12.0      
6                                               11.0      
7                                               10.0      
8                                                7.0      
9                                                6.0      

   Number of cases fulfilling the reporting criteria  Number of death cases  \
0                                               38.0                    0.0   
1                                               48.0                    0.0   
2                                               54.0                    0.0   
3                                               61.0                    0.0   
4                                               67.0                    0.0   
5                                               68.0                    0.0   
6                                               71.0                    0.0   
7                                               76.0                    0.0   
8                                               78.0                    0.0   
9                                               81.0                    0.0   

   Number of discharge cases  Number of probable cases  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   
5                        0.0                       0.0   
6                        0.0                       0.0   
7                        0.0                       0.0   
8                        0.0                       0.0   
9                        0.0                       0.0   

   Number of hospitalised cases in critical condition  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    
5                                                NaN    
6                                                NaN    
7                                                NaN    
8                                                NaN    
9                                                NaN    

   Number of cases tested positive for SARS-CoV-2 virus by nucleic acid tests  \
0                                                NaN                            
1                                                NaN                            
2             

In [5]:
# Basic information of dataframe
covid_hk_case_count_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1238 entries, 0 to 1237
Data columns (total 14 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   As of date                                                                   1238 non-null   object 
 1   As of time                                                                   194 non-null    object 
 2   Number of confirmed cases                                                    724 non-null    float64
 3   Number of ruled out cases                                                    89 non-null     float64
 4   Number of cases still hospitalised for investigation                         89 non-null     float64
 5   Number of cases fulfilling the reporting criteria                            89 non-null     float64
 6   Number of death cases                   

# Modify Data Types

In [6]:
# Modify data type for datatime column
covid_hk_case_count_df['As of date'] = pd.to_datetime(covid_hk_case_count_df['As of date'], format='%d/%m/%Y')

# Exploratory Data Analysis (EDA)

## Univariate Analysis

## Field - As of date

In [7]:
# Report date range of data set
min_date, max_date = covid_hk_case_count_df['As of date'].agg(['min', 'max']).astype('str').to_list()
print('Date range of as of date:')
print(f'Min date: {min_date}')
print(f'Max date: {max_date}')

Date range of as of date:
Min date: 2020-01-08
Max date: 2023-05-29


In [8]:
# Dates which are not included in the dataset
covid_hk_case_count_df = covid_hk_case_count_df.set_index('As of date', drop=False)
missing_date_index = pd.date_range(start=min_date, end=max_date).difference(covid_hk_case_count_df.index).astype('str')

if len(missing_date_index) == 0:
    print('There is no missing date in the dataset.')
else:
    print(f'There are missing dates in the dataset, between {missing_date_index.min()} and {missing_date_index.max()}.')

There is no missing date in the dataset.


In [9]:
print('All values are distinct:', covid_hk_case_count_df['As of date'].nunique() == covid_hk_case_count_df['As of date'].shape[0])

All values are distinct: True


## Field - As of time

In [10]:
# Distinct values in as of time column
print('Distinct values in as of time column are:',
      str(covid_hk_case_count_df['As of time'].drop_duplicates().to_list())[1:-1])

Distinct values in as of time column are: '12:00', '23:59', '16:00', nan


## Fields - Columns with Counts 

In [11]:
covid_module.print_cumulative_columns(covid_hk_case_count_df)

The cumulative columns in dataset are:
Number of ruled out cases
Number of cases fulfilling the reporting criteria
Number of death cases
Number of cases tested positive for SARS-CoV-2 virus by nucleic acid tests
Number of cases tested positive for SARS-CoV-2 virus by rapid antigen tests

The non-cumulative columns in dataset are:
Number of confirmed cases
Number of cases still hospitalised for investigation
Number of discharge cases
Number of probable cases
Number of hospitalised cases in critical condition
Number of positive nucleic acid test laboratory detections
Number of death cases related to COVID-19


The columns 'Number of confirmed cases' and 'Number of discharge cases' are cumulative columns indeed. <br>
There are only 2 dates (rows) that these 2 columns are not monotonic. 

For 'Number of confirmed cases', <br>
The values on 2021-09-02 to 2021-09-04 12113, 12112 and 12113 <br>
The values on 2021-10-08 to 2021-10-10 are 12253, 12230 and 12261

For 'Number of discharge cases', <br>
The values on 2021-09-02 to 2021-09-04 11799, 11798 and 11802 <br>
The values on 2021-10-08 to 2021-10-10 are 11945, 11926 and 11947

Based on the following observations, it is likely that the values on 2021-09-03 and 2021-10-09 are not accurate. <br>
1. There are a total of 724 non-null data points in the dataset, but only 2 of these data points (for the dates 2021-09-03 and 2021-10-09) exhibit non-monotonic behavior.
2. The value on 2021-09-04 is larger than the value on 2021-09-02. Similarly, the value on 2021-10-10 is larger than the value on 2021-10-08.

## Multivariate Analysis

## Fields - As of Date and Number of confirmed cases

In [16]:
# # Line chart - Cumulative Numbers of Confirmed Cases (Hong Kong)
# plt.subplots(figsize=(15, 8))
# plt.plot(covid_hk_case_count_df.index, covid_hk_case_count_df['Number of confirmed cases'])
# plt.title('Cumulative Number of Confirmed Covid-19 Cases (Hong Kong)')
# plt.xlabel('Year-Month')
# plt.ylabel('Number of Confirmed Cases')
# plt.xticks(rotation=90)
# plt.show()

## Range of Data

In [35]:
# Range of data in each column
for target_col in covid_hk_case_count_df.columns:
    if target_col not in ['As of date', 'As of time']:
        covid_module.print_range_of_data(covid_hk_case_count_df, 'As of date', target_col)

The column "Number of confirmed cases" has data between:
From 2020-01-08 to 2021-12-31

The column "Number of ruled out cases" has data between:
From 2020-01-08 to 2020-04-05

The column "Number of cases still hospitalised for investigation" has data between:
From 2020-01-08 to 2020-04-05

The column "Number of cases fulfilling the reporting criteria" has data between:
From 2020-01-08 to 2020-04-05

The column "Number of death cases" has data between:
From 2020-01-08 to 2023-01-29

The column "Number of discharge cases" has data between:
From 2020-01-08 to 2022-02-12
From 2022-04-26 to 2022-12-04

The column "Number of probable cases" has data between:
From 2020-01-08 to 2021-12-31

The column "Number of hospitalised cases in critical condition" has data between:
From 2020-04-06 to 2023-01-29

The column "Number of cases tested positive for SARS-CoV-2 virus by nucleic acid tests" has data between:
From 2022-01-01 to 2023-01-29

The column "Number of cases tested positive for SARS-CoV-2

## Missing Value Analysis

In [14]:
# Number of missing values in each column
covid_module.print_missing_val_count(covid_hk_case_count_df)

The following columns have missing values:
As of time: 1044 (84.3%)
Number of confirmed cases: 514 (41.5%)
Number of ruled out cases: 1149 (92.8%)
Number of cases still hospitalised for investigation: 1149 (92.8%)
Number of cases fulfilling the reporting criteria: 1149 (92.8%)
Number of death cases: 120 (9.7%)
Number of discharge cases: 248 (20.0%)
Number of probable cases: 514 (41.5%)
Number of hospitalised cases in critical condition: 209 (16.9%)
Number of cases tested positive for SARS-CoV-2 virus by nucleic acid tests: 844 (68.2%)
Number of cases tested positive for SARS-CoV-2 virus by rapid antigen tests: 900 (72.7%)
Number of positive nucleic acid test laboratory detections: 1118 (90.3%)
Number of death cases related to COVID-19: 1118 (90.3%)
